src/rgw/rgw_auth.h

```diff
@@ -0,0 +1,340 @@
+  virtual ~RGWAuthApplier() {};
+
+  /* Fill provided RGWUserInfo with information about the account that
+   * RGWOp will operate on. Errors are handled solely through exceptions.
+   *
+   * XXX: be aware that the "account" term refers to rgw_user. The naming
+   * is legacy. */
+  virtual void load_acct_info(RGWUserInfo& user_info) const = 0; /* out */
+
+  /* Apply any changes to request state. This method will be most useful for
+   * TempURL of Swift API or AWSv4. */
+  virtual void modify_request_state(req_state * s) const {}      /* in/out */
+};
+
+
+/* RGWRemoteAuthApplier targets those authentication engines which don't need
+ * to ask the RADOS store while performing the auth process. Instead, they
+ * obtain credentials from an external source like Keystone or LDAP.
+ *
+ * As the authenticated user may not have an account yet, RGWRemoteAuthApplier
+ * must be able to create it basing on data passed by an auth engine. Those
+ * data will be used to fill RGWUserInfo structure. */
+class RGWRemoteAuthApplier : public RGWAuthApplier {
+public:
+  class AuthInfo {
+    friend class RGWRemoteAuthApplier;
+  protected:
+    const rgw_user acct_user;
+    const std::string acct_name;
+    const uint32_t perm_mask;
+    const bool is_admin;
+
+  public:
+    AuthInfo(const rgw_user acct_user,
+             const std::string acct_name,
+             const uint32_t perm_mask,
+             const bool is_admin)
+    : acct_user(acct_user),
+      acct_name(acct_name),
+      perm_mask(perm_mask),
+      is_admin(is_admin) {
+    }
+  };
+
+  using aclspec_t = RGWIdentityApplier::aclspec_t;
+  typedef std::function<uint32_t(const aclspec_t&)> acl_strategy_t;
+
+protected:
+  /* Read-write is intensional here due to RGWUserInfo creation process. */
+  RGWRados * const store;
+
+  /* Supplemental strategy for extracting permissions from ACLs. Its results
+   * will be combined (ORed) with a default strategy that is responsible for
+   * handling backward compatibility. */
+  const acl_strategy_t extra_acl_strategy;
+
+  const AuthInfo info;
+
+  virtual void create_account(const rgw_user acct_user,
+                              RGWUserInfo& user_info) const;          /* out */
+
+public:
+  RGWRemoteAuthApplier(CephContext * const cct,
+                       RGWRados * const store,
+                       acl_strategy_t&& extra_acl_strategy,
+                       const AuthInfo info)
+    : RGWAuthApplier(cct),
+      store(store),
+      extra_acl_strategy(std::move(extra_acl_strategy)),
+      info(info) {
+  }
+
+  virtual uint32_t get_perms_from_aclspec(const aclspec_t& aclspec) const override;
+  virtual bool is_admin_of(const rgw_user& uid) const override;
+  virtual bool is_owner_of(const rgw_user& uid) const override;
+  virtual uint32_t get_perm_mask() const { return info.perm_mask; }
+  virtual std::string to_str() const override;
+  virtual void load_acct_info(RGWUserInfo& user_info) const override; /* out */
```

Comment:
`override` is enough to document that these are virtual functions, so the `virtual` part isn't necessary